### PR TITLE
Door rewrite 20230301

### PR DESCRIPTION
**Pull request by shinespeciall**
use a new LevelDoorVector class to replace the old Door class since the old one is hard to maintain, and impossible to implement undo and redo features. now all the logic relates to the old Door class got an overhaul.

### CODE REVIEW

File: LevelComponents/Level.cpp [hunk 1]
```diff
@@ -327,71 +300,21 @@ namespace LevelComponents
...
-        std::vector<Door *> roomDoors;
-        // Distribute door data
-        for (unsigned int i = 0; i < doors.size(); ++i)
-        {
-            if (doors[i]->GetRoomID() == (int) roomId)
+            else
             {
-                Door *newDoor = new Door(*doors[i]);
-                roomDoors.push_back(newDoor);
+                rooms[i]->SetCurrentEntitySet(37);
```
Comment (Goldensunboy) [inline LevelComponents/Level.cpp:315]: What is 37?

File: LevelComponents/LevelDoorVector.cpp [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+#include <QRegularExpression>
+
+#include "ROMUtils.h"
+
+LevelComponents::LevelDoorVector::LevelDoorVector(unsigned int doorDataStartAddr)
+{
+    unsigned char *doorPtr = ROMUtils::ROMFileMetadata->ROMDataPtr + doorDataStartAddr;
+    int currentDoorGlobalId = 0;
+
+    // this build cannot work in your computer
```
Comment (Goldensunboy) [inline LevelComponents/LevelDoorVector.cpp:12]: What does this mean

File: LevelComponents/Level.cpp (hunk 1 above)
Comment (shinespeciall) [reply LevelComponents/Level.cpp]: some default id for entityset, i moved it into the class as a init value for the variable

File: LevelComponents/LevelDoorVector.cpp (hunk 2 above)
Comment (shinespeciall) [reply LevelComponents/LevelDoorVector.cpp]: deleted

**Review by Goldensunboy — APPROVED** [review]

**Pull request merged**